File: java/app-encryption/src/main/java/com/godaddy/asherah/crypto/envelope/AeadEnvelopeCrypto.java
package com.godaddy.asherah.crypto.envelope;

import static com.godaddy.asherah.crypto.bufferutils.ManagedBufferUtils.wipeByteArray;

import java.time.Instant;

import com.godaddy.asherah.crypto.AeadCrypto;
import com.godaddy.asherah.crypto.keys.CryptoKey;

public abstract class AeadEnvelopeCrypto extends AeadCrypto {

  /**
   * Encrypts a {@link CryptoKey} with another {@link CryptoKey}.
   * @param key The key to encrypt.
   * @param keyEncryptionKey The key encryption key.
   * @return The encrypted key bytes.
   */
  public byte[] encryptKey(final CryptoKey key, final CryptoKey keyEncryptionKey) {
    return key.withKey(keyBytes -> {
      return encrypt(keyBytes, keyEncryptionKey);
    });
  }

  /**
   * Decrypts an encrypted key.
   * @param encryptedKey The encrypted key bytes.
   * @param encryptedKeyCreated The creation time of the encrypted key.
   * @param keyEncryptionKey The key encryption key.
   * @return A decrypted {@link CryptoKey} object.
   */
  public CryptoKey decryptKey(final byte[] encryptedKey, final Instant encryptedKeyCreated,
      final CryptoKey keyEncryptionKey) {
    return decryptKey(encryptedKey, encryptedKeyCreated, keyEncryptionKey, false);
  }

  /**
   * Decrypts an encrypted key.
   * @param encryptedKey The encrypted key bytes.
   * @param encryptedKeyCreated The creation time of the encrypted key.
   * @param keyEncryptionKey The key encryption key.
   * @param revoked The revocation status of the key.
   * @return A decrypted {@link CryptoKey} object.
   */
  public CryptoKey decryptKey(final byte[] encryptedKey, final Instant encryptedKeyCreated,
      final CryptoKey keyEncryptionKey, final boolean revoked) {
    //Invert this into functional
    byte[] decryptedKey = decrypt(encryptedKey, keyEncryptionKey);
    try {
      return generateKeyFromBytes(decryptedKey, encryptedKeyCreated, revoked);
    }
    finally {
      wipeByteArray(decryptedKey);
    }
  }

  /**
   * Encrypts the payload and the key to create the data row record.
   * @param plainText The payload to be encrypted.
   * @param keyEncryptionKey The key encryption key.
   * @return An {@link EnvelopeEncryptResult} object/data row record (DRR).
   */
  public EnvelopeEncryptResult envelopeEncrypt(final byte[] plainText, final CryptoKey keyEncryptionKey) {
    return envelopeEncrypt(plainText, keyEncryptionKey, null);
  }


  /**
   * Encrypts the payload and the key to create the data row record.
   * @param plainText The payload to be encrypted.
   * @param keyEncryptionKey The key encryption key.
   * @param userState The KeyMeta for the {@code keyEncryptionKey}.
   * @return An {@link EnvelopeEncryptResult} object/data row record (DRR).
   */
  public EnvelopeEncryptResult envelopeEncrypt(final byte[] plainText, final CryptoKey keyEncryptionKey,
      final Object userState) {
    try (CryptoKey dataEncryptionKey = generateKey()) {
      EnvelopeEncryptResult result = new EnvelopeEncryptResult();
      result.setCipherText(encrypt(plainText, dataEncryptionKey));
      result.setEncryptedKey(encryptKey(dataEncryptionKey, keyEncryptionKey));
      result.setUserState(userState);
      return result;
    }
  }

  /**
   * Decrypts the encryptedKey and then uses the decrypted key to decrypt the encrypted payload in the data row record.
   * @param cipherText The encrypted payload.
   * @param encryptedKey The encrypted key.
   * @param keyCreated The creation time of the data row record.
   * @param keyEncryptionKey The key encryption key.
   * @return The decrypted payload.
   */
  public byte[] envelopeDecrypt(final byte[] cipherText, final byte[] encryptedKey,
      final Instant keyCreated, final CryptoKey keyEncryptionKey) {
    try (CryptoKey plaintextKey = decryptKey(encryptedKey, keyCreated, keyEncryptionKey)) {
      return decrypt(cipherText, plaintextKey);
    }
  }
}

File: csharp/AppEncryption/Crypto/Envelope/AeadEnvelopeCrypto.cs
using System;
using System.Runtime.CompilerServices;
using GoDaddy.Asherah.Crypto.BufferUtils;
using GoDaddy.Asherah.Crypto.Keys;

[assembly: InternalsVisibleTo("DynamicProxyGenAssembly2")]

namespace GoDaddy.Asherah.Crypto.Envelope
{
  public abstract class AeadEnvelopeCrypto : AeadCrypto
  {
    /// <summary>
    /// Encrypts a <see cref="CryptoKey"/> with another <see cref="CryptoKey"/>.
    /// </summary>
    ///
    /// <param name="key">The key to encrypt.</param>
    /// <param name="keyEncryptionKey">The key encryption key.</param>
    /// <returns>The encrypted key bytes.</returns>
    public virtual byte[] EncryptKey(CryptoKey key, CryptoKey keyEncryptionKey)
    {
      return key.WithKey(keyBytes => Encrypt(keyBytes, keyEncryptionKey));
    }

    /// <summary>
    /// Decrypts an encrypted key.
    /// </summary>
    ///
    /// <param name="encryptedKey">The encrypted key bytes.</param>
    /// <param name="encryptedKeyCreated">The creation time of the encrypted key.</param>
    /// <param name="keyEncryptionKey">The key encryption key.</param>
    /// <returns>A decrypted <see cref="CryptoKey"/> object.</returns>
    public virtual CryptoKey DecryptKey(byte[] encryptedKey, DateTimeOffset encryptedKeyCreated, CryptoKey keyEncryptionKey)
    {
      return DecryptKey(encryptedKey, encryptedKeyCreated, keyEncryptionKey, false);
    }

    /// <summary>
    /// Decrypts an encrypted key.
    /// </summary>
    ///
    /// <param name="encryptedKey">The encrypted key bytes.</param>
    /// <param name="encryptedKeyCreated">The creation time of the encrypted key.</param>
    /// <param name="keyEncryptionKey">The key encryption key.</param>
    /// <param name="revoked">The revocation status of the key.</param>
    /// <returns>A decrypted <see cref="CryptoKey"/> object.</returns>
    public virtual CryptoKey DecryptKey(
        byte[] encryptedKey, DateTimeOffset encryptedKeyCreated, CryptoKey keyEncryptionKey, bool revoked)
    {
      byte[] decryptedKey = Decrypt(encryptedKey, keyEncryptionKey);
      try
      {
        return GenerateKeyFromBytes(decryptedKey, encryptedKeyCreated, revoked);
      }
      finally
      {
        ManagedBufferUtils.WipeByteArray(decryptedKey);
      }
    }

    /// <summary>
    /// Encrypts the payload and the key to create the data row record.
    /// </summary>
    ///
    /// <param name="plainText">The payload to be encrypted.</param>
    /// <param name="keyEncryptionKey">The key encryption key.</param>
    /// <returns>A <see cref="EnvelopeEncryptResult"/>object/data row record (DRR).</returns>
    public virtual EnvelopeEncryptResult EnvelopeEncrypt(byte[] plainText, CryptoKey keyEncryptionKey)
    {
      return EnvelopeEncrypt(plainText, keyEncryptionKey, null);
    }

    /// <summary>
    /// Encrypts the payload and the key to create the data row record.
    /// </summary>
    ///
    /// <param name="plainText">The payload to be encrypted.</param>
    /// <param name="keyEncryptionKey">The key encryption key.</param>
    /// <param name="userState">The KeyMeta for the <see cref="keyEncryptionKey"/>.</param>
    /// <returns>A <see cref="EnvelopeEncryptResult"/>object/data row record (DRR).</returns>
    public virtual EnvelopeEncryptResult EnvelopeEncrypt(byte[] plainText, CryptoKey keyEncryptionKey, object userState)
    {
      using (CryptoKey dataEncryptionKey = GenerateKey())
      {
        EnvelopeEncryptResult result = new EnvelopeEncryptResult
        {
          CipherText = Encrypt(plainText, dataEncryptionKey),
          EncryptedKey = EncryptKey(dataEncryptionKey, keyEncryptionKey),
          UserState = userState,
        };

        return result;
      }
    }

    /// <summary>
    /// Decrypts the encryptedKey and then uses the decrypted key to decrypt the encrypted payload in the data row record.
    /// </summary>
    ///
    /// <param name="cipherText">The encrypted payload.</param>
    /// <param name="encryptedKey">The encrypted key.</param>
    /// <param name="keyCreated">The creation time of the data row record.</param>
    /// <param name="keyEncryptionKey">The key encryption key.</param>
    /// <returns>The decrypted payload.</returns>
    public virtual byte[] EnvelopeDecrypt(byte[] cipherText, byte[] encryptedKey, DateTimeOffset keyCreated, CryptoKey keyEncryptionKey)
    {
      using (CryptoKey plaintextKey = DecryptKey(encryptedKey, keyCreated, keyEncryptionKey))
      {
        return Decrypt(cipherText, plaintextKey);
      }
    }
  }
}
